# Практическое задание №1 по теме "Случайные события. Условная вероятность. Формула Байеса.  Независимые испытания".

In [1]:
import itertools
import numpy as np

def arrangements(n, k):
    return np.math.factorial(n) // np.math.factorial(n - k)

def combinations(n, k):
    return np.math.factorial(n) // (np.math.factorial(k) * np.math.factorial(n - k))

def permutations(n):
    return np.math.factorial(n)

__Задача 1__

Группа студентов изучает 10 различных дисциплин. Сколькими способами можно составить расписание на понедельник, если в этот день должно быть 4 разных занятия?

___Решение___

а) Если расписание начала занятий в понедельник у группы студентов значения не имеет, значит можно не учитывать различные варицаии в каком порядке какой предмет идет, а лишь учитывать какие конкретно предметы будут в понедельник.

Воспользуемся формулой для сочетания: $$C_n^k = \frac{n!}{k!(n - k)!}$$

In [2]:
n = 10
k = 4
res = combinations(n, k)

print(f"Количество комбинаций: {res}")

Количество комбинаций: 210


б) Если все же учитывать расписание занятий, их порядок. То есть первая пара, например, матан, а вторая физика, тогда нужно учитывать все варианты.

Воспользуемся формулой для размещения: $$C_n^k = \frac{n!}{(n - k)!}$$

In [3]:
n = 10
k = 4
res = arrangements(n, k)

print(f"Количество размещений: {res}")

Количество размещений: 5040


___Ответ:___ 210 способами - если расписание (порядок) предметов в понедельник не имеет значение; 5040 способов - если расписание имеет значение.

---

__Задача 2__

Из колоды в 52 карты вынимают случайным образом 4 карты. Найти число исходов, соответствующих тому, что был вытянут хотя бы один туз.

___Решение___

Чтобы решить эту задачу, нужно сначала рассмотреть отдельно ситуации, когда попались или один, или два, или три, или четыре туза из выбранных четырех карт.

1. Попалось 4 туза

Ситуация, в которой попалось бы все четыре туза всего 1.

$$N_1 = 1$$

In [4]:
n_1 = 1
n_1

1

2. Попалось 3 туза

Число сочетаний трех тузов из четырех: $$C_{4}^{3} = \frac{4!}{(4 - 3)!}$$

Оставшаяся карта выбирается из 48 карт  — тузы не рассматриваются, так как они уже выбраны: $$C_{48}^{1}= \frac{48!}{(48 - 1)!}$$

Число сочетаний, когда из колоды, в которой 52 карт, выбраны четыре, три из которых — тузы: $$N_2 = C_{4}^{3} \cdot C_{48}^{1} = 192$$

In [5]:
n_2 = combinations(4, 3) * combinations(48, 1)
n_2

192

3. Попалось 2 туза

Число сочетаний двух тузов из четырех: $$ C_{4}^{2} = \frac{4!}{(4 - 2)!}$$

Оставшиеся две карты выбираются из 48 карт  — тузы не рассматриваются, так как они уже выбраны: $$ C_{48}^{2}= \frac{48!}{(48 - 2)!}$$

Число сочетаний, когда из колоды, в которой 52 карты, выбраны четыре, две из которых — тузы: $$N_3 = C_{4}^{2} \cdot C_{48}^{2} = 6768$$

In [6]:
n_3 = combinations(4, 2) * combinations(48, 2)
n_3

6768

4. Попался 1 туз

Число сочетаний одного туза из четырех: $$ C_{4}^{1} = \frac{4!}{(4 - 1)!}$$

Оставшиеся три карты выбираются из 48 карт  — тузы не рассматриваются, так как они уже выбраны: $$ C_{48}^{3}= \frac{48!}{(48 - 3)!}$$

Число сочетаний, когда из колоды, в которой 52 карт, выбраны четыре, одна из которых — туз: $$N_4 = C_{4}^{1} \cdot C_{48}^{3} = 69184$$

In [7]:
n_4 = combinations(4, 1) * combinations(48, 3)
n_4

69184

Осталось сложить полученные числа сочетаний всех случаев: $$N_1 + N_2 + N_3 + N_4 = 76145$$

In [8]:
result = n_1 + n_2 + n_3 + n_4
result 

76145

___Ответ:___ 76145 исходов

---

__Задача 3__

Семь человек рассаживаются наудачу на скамейке. Какова вероятность того, что два определённых человека будут сидеть рядом?

___Решение___

Пусть $A$ - событие, когда два определенных человека сидят рядом, $m$ - число событий, в результате которых произошло событие $A$, $n$ - общее число испытаний, тогда

$$n = P_7 = 7! = 5040$$

Теперь подсчитаем количество событий $m$, когда два определенных человека сидят рядом.

Пусть буквы - это обозначение людей на скамейке, причем $A$ и $B$ - те самые два определенных человека, тогда можно представить следующие расположения:

$$ABcdefg$$
$$cABdefg$$
$$cdABefg$$
$$cdeABfg$$
$$cdefABg$$
$$cdefgAB$$

Получается $6$ случаев, если мы переставим местами $A$ и $B$, то прибавится еще $6$ случаев. Итого $12$. 

Так как нам важен порядок, нужно учесть расположения остальных пяти человек, то есть посчитать количество их перестановок.

Количество перестановок для пяти человек для каждого из 12 вариантов: $P_5 = 5!$

Количество случаев возникновения события $A$:

$$m = 12 \cdot P_5 = 12 \cdot 5!=  1440$$

$$P(A) = \frac{m}{n} = \frac{1440}{5040} = \frac{2}{7} \approx 0.2857$$

In [9]:
n = permutations(7)
n

5040

In [10]:
m = 12 * permutations(5)
m

1440

In [11]:
P = m/n
np.round(P, 4)

0.2857

Проверка полученного результата:

In [12]:
from pprint import pprint

count = 0
result = []
for p in itertools.permutations('0123456', 7):
    st = ''.join(p)
    if st.find("01") > -1 or st.find("10") > -1 :
        result.append(st)
    count += 1
print(f"n: {count}")
print(f"m: {len(result)}")
print(f"P(A): {np.round(m/n, 4)}")

n: 5040
m: 1440
P(A): 0.2857


___Ответ:___ $\frac{2}{7}$

---

__Задача 4__

Из 60 вопросов, входящих в экзаменационные билеты, студент знает 50. Какова вероятность того, что среди трёх наугад выбранных вопросов студент знает: а) все? б) два?

___Решение___

Порядок в данной задаче значения не имеет. 

___а)___ 

Пусть $A$ - событие, когда среди трёх наугад выбранных вопросов студент знает все, $m$ - число всех случаев, когда произошло событие $A$, $n$ - общее число случаев, тогда


$$m = C_{50}^{3}$$

$$n = C_{60}^{3}$$

$$P(A) = \frac{m}{n} = \frac{C_{50}^{3}}{C_{60}^{3}} = \frac{50!}{3!\,47!} \cdot \frac{3!\,57!}{60!} = \frac{48\cdot 49\cdot 50}{58\cdot 59\cdot 60}\approx 0.5728$$

In [13]:
n = combinations(60,3)
m = combinations(50,3)
p = np.round(m/n, 4)
p

0.5728

___б)___ 

Пусть $B$ - событие, когда среди трёх наугад выбранных вопросов студент знает только 2, $m$ - число всех случаев, когда произошло событие $B$, $n$ - общее число случаев.

$$n = C_{60}^{3}$$

Рассмотрим ситуацию, когда выбирается 2 знакомых студенту вопроса из 50, тогда число сочетаний будет равно:

$$m_1 = C_{50}^{2}$$

А третий вопрос (не из тех 50) выбирается из 10 вопросов, которые студент не знает. Таких случаев:

$$m_2 = C_{10}^{1}$$

И тогда число сочетаний, что среди трёх наугад выбранных вопросов студент знает два, а третий не знает:

$$m = m_1 \cdot m_2 = C_{50}^{2} \cdot C_{10}^{1}$$

$$P(B) = \frac{m}{n} = \frac{C_{50}^{2}\cdot C_{10}^{1}}{C_{60}^{3}} = \frac{50!}{2!\,48!} \cdot \frac{10!}{1!\,9!} \cdot \frac{3!\,57!}{60!} = \frac{49\cdot 50 \cdot 10 \cdot 3}{58\cdot 59\cdot 60}\approx 0.358$$

In [14]:
n = combinations(60,3)
m = combinations(50,2) * combinations(10, 1)
p = np.round(m/n, 4)
p

0.358

___Ответ:___ 

___a)___  $P(A) \approx 0.5728$, <br>
___б)___ $P(B) \approx 0.358$

---

__Задача 5__

Бросается игральная кость. Пусть событие `A` - появление чётного числа, событие `B` - появление числа больше трёх. Являются ли эти события независимыми?

___Решение___

Два события называются __совместными__, если появление одного из них не исключает появления другого в одном и том же испытании.

События $A$ и $B$ являются совместными событиями, так как при выпадении 4 или 6, например, одновременно происходят оба эти события.

Событие $B$ называется __независимым__ от события $A$, если появление события $A$ не изменяет вероятности $B$, то есть $P(B|A) = P(B)$

$$P(B | A) = \dfrac{P(AB)}{P(A)}$$

$$P(A) = \frac{1}{2}$$

$$P(B) = \frac{1}{2}$$

Только 2 из 6 событий соответствуют наступлению событий  $A$  и  $B$  одновременно, значит

$$P(AB) = \frac{1}{3} $$

Таким образом, условная вероятность $P(AB)$:

$$P(B | A) = \dfrac{P(AB)}{P(A)} = \frac{1}{3} : \frac{1}{2} = \frac{2}{3}$$


Получилось, что равенство $P(B|A) = P(B)$ неверно $\Bigl(\frac{2}{3} \neq \frac{1}{2}\Bigr)$, события не являются независимыми.

___Ответ:___ Нет, события не являются независимыми!

---

__Задача 6__

Допустим, имеется некоторая очень редкая болезнь (поражает 0.1 % населения). Вы приходите к врачу, вам делают тест на эту болезнь, и тест оказывается положительным. Врач говорит вам, что этот тест верно выявляет 99 % больных этой болезнью и всего лишь в 1 % случаев даёт ложный положительный ответ.

_Вопрос_: какова вероятность, что вы действительно больны ей?

_Подсказка_: вновь используйте формулу Байеса с раскрытием знаменателя с помощью формулы полной вероятности.

___Решение___

Сразу же определимся с тем, какие события мы здесь рассматриваем. Пусть событие $A$ заключается в том, что тест оказался положительным (+). Кроме того, пусть событие $B_1$ заключается в том, что мы все-таки больны болезнью, а событие $B_2$ означает, что мы здоровы.

Нам нужно определить какова вероятность, что мы действительно больны при условии, что наш тест положительный (+), то есть $P(B_1|A)$ - ?

По условию болезнь является очень редкой и поражает 0.1% населения, что означает следующее:

$$P(B_1) = 0.001$$

$$P(B_2) = 1 - P(B_1) = 0.999$$

Так как тест верно выявляет 99% больных, то 

$$P(A|B_1) = 0.99$$

Так как в случае здоровых людей тест показывает 1% ложно-положительных результатов, то 

$$P(A|B_2) = 0.01$$

По формуле Байеса $P(B_1|A)$:

$$P(B_1|A) = \frac{P(A|B_1) \cdot P(B_1)}{P(A)}$$

Найдем $P(A)$ с помощью формулы полной вероятности:

$$P(A) = P(B_1) \cdot P(A | B_1) + P(B_2) \cdot P(A | B_2) = 0.001\cdot 0.99 + 0.999\cdot 0.01 = 0.01098$$

In [15]:
p_a = 0.001*0.99 + 0.999*0.01
p_a

0.01098

Подставляем найденные значение в формулу Байеса:

$$P(B_1|A) = \frac{P(A|B_1) \cdot P(B_1)}{P(A)} = \frac{0.99\cdot 0.001}{0.01098} \approx  0.0902$$

In [16]:
p_b1_a = 0.001*0.99/p_a
np.round(p_b1_a, 4)

0.0902

___Ответ:___ Вероятность болезни $\approx  0.0902$

---